### PR TITLE
Add Normalizer interface for nested parameter normalization

### LINKED ISSUE #491: Support nested parameter alias normalization for Amadeus book_flight traveler fields
## Background

As part of the parameter alias normalization work (PR #492), we audited all connector actions for places where agents might send different parameter names than what the schema expects.

The `ParameterAliaser` interface added in that PR handles **top-level** parameter key renaming. It cannot handle nested objects or arrays.

## The Problem

The `amadeus.book_flight` action requires a `travelers` array where each element uses **camelCase** field names:

```json
{
  "travelers": [
    {
      "name": { "firstName": "John", "lastName": "Doe" },
      "dateOfBirth": "1990-01-15",
      "gender": "MALE",
      "contact": { "email": "john@example.com", "phone": "+1234567890" }
    }
  ]
}
```

This is inconsistent with the **snake_case** convention used by every other connector. An agent following the system's established convention would naturally send `first_name`, `last_name`, `date_of_birth` — and fail silently after the user approves.

## Design

Add a separate optional `Normalizer` interface alongside `ParameterAliaser`:

```go
// Normalizer is an optional interface for actions that need arbitrary
// parameter transformation beyond flat key aliasing. The API layer calls
// Normalize() after ParameterAliaser (if both are implemented), before
// storing the approval or executing the action.
type Normalizer interface {
    Normalize(params json.RawMessage) json.RawMessage
}
```

The API layer adds one more type assertion at the same two call sites as `ParameterAliaser` (`POST /approvals/request` and `POST /actions/execute`). Each connector handles its own traversal logic — no path syntax to invent or parse.

`amadeus.book_flight` would implement `Normalize()` to walk the `travelers` array and rewrite snake_case traveler fields to camelCase.

This approach also naturally handles any future connector with unusual nested structures.

## What Needs To Be Done

- [ ] Add `Normalizer` interface to `connectors/connector.go`
- [ ] Wire type assertion for `Normalizer` in `api/agent_approvals.go` (after `ParameterAliaser` normalization)
- [ ] Wire type assertion for `Normalizer` in `api/action_execute.go` (same position)
- [ ] Implement `Normalize()` on `amadeus.book_flight` to rewrite `first_name`→`firstName`, `last_name`→`lastName`, `date_of_birth`→`dateOfBirth` in each traveler object
- [ ] Add tests for the Amadeus normalization

## Acceptance Criteria

- [ ] Agents sending `first_name`/`last_name`/`date_of_birth` in traveler objects have them normalized to `firstName`/`lastName`/`dateOfBirth` before storage and execution
- [ ] Agents sending the correct camelCase names continue to work unchanged
- [ ] The `Normalizer` interface is documented as the pattern for any connector needing non-flat parameter transformation

Relates to #492 (flat alias normalization), #493 (pre-ingestion schema validation).

### PR DESCRIPTION
**Pull request by chiedoclaude**
## Summary

- Adds a `Normalizer` interface to `connectors/connector.go` for actions that need arbitrary nested parameter transformation beyond `ParameterAliaser`'s flat key→key map
- Wires `Normalizer` type assertion in both `api/agent_approvals.go` and `api/action_execute.go`, running after `ParameterAliaser` normalization
- Implements `Normalize()` on `amadeus.book_flight` to rewrite snake_case traveler fields (`first_name`→`firstName`, `last_name`→`lastName`, `date_of_birth`→`dateOfBirth`) inside the `travelers` array

Closes #491

### Claude Code
- [x] Unit tests for snake_case → camelCase normalization
- [x] Test that already-camelCase fields pass through unchanged
- [x] Test that camelCase takes precedence when both forms are present
- [x] Test multiple travelers (mix of snake_case and camelCase)
- [x] Test edge cases (no travelers, empty params, invalid JSON)
- [x] Compile-time interface assertion (`var _ connectors.Normalizer = (*bookFlightAction)(nil)`)
- [x] Verify existing amadeus tests still pass

### OpenClaw
- [ ] Verify end-to-end with a real agent sending snake_case traveler fields
- [ ] Confirm camelCase-only agents are unaffected in production
- [ ] Review whether other connectors need similar normalization

https://claude.ai/code/session_01BK6XBcXw9LUBLNbeXsf2t8

### CODE REVIEW

File: connectors/amadeus/book_flight_normalize.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	for snake, camel := range snakeToCamelTravelerFields {
+		if _, hasCamel := t[camel]; hasCamel {
+			continue // canonical already present
+		}
+		if val, hasSnake := t[snake]; hasSnake {
+			t[camel] = val
+			delete(t, snake)
+			changed = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/amadeus/book_flight_normalize.go]: Fixed in 603233a. Now deletes the snake_case key even when the camelCase canonical is already present, preventing duplicate fields from reaching the Amadeus API. Also added assertions in the CamelCaseTakesPrecedence test to catch this going forward.

File: connectors/amadeus/book_flight_normalize.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			nameChanged := false
+			for snake, camel := range snakeToCamelTravelerFields {
+				if _, hasCamel := name[camel]; hasCamel {
+					continue
+				}
+				if val, hasSnake := name[snake]; hasSnake {
+					name[camel] = val
+					delete(name, snake)
+					nameChanged = true
+				}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/amadeus/book_flight_normalize.go]: Fixed in the same commit (603233a). Applied the identical delete-before-continue pattern to the name sub-object block.

File: api/action_execute.go [hunk 3]
```diff
@@ -115,6 +115,16 @@ func handleStandingApprovalPath(w http.ResponseWriter, r *http.Request, deps *De
...
 
+	// ── Normalize nested parameter structures ────────────────────
+
+	if deps.Connectors != nil {
+		if action, ok := deps.Connectors.GetAction(req.Action.Type); ok {
+			if normalizer, ok := action.(connectors.Normalizer); ok {
+				params = normalizer.Normalize(params)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/action_execute.go]: Good catch. Merged both normalization steps under a single GetAction lookup in 603233a.

File: connectors/amadeus/book_flight_normalize_test.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+	"testing"
+
+	"github.com/supersuit-tech/permission-slip-web/connectors"
+)
+
+// Verify bookFlightAction implements Normalizer at compile time.
+var _ connectors.Normalizer = (*bookFlightAction)(nil)
+
+func TestBookFlightNormalize_SnakeCaseToCalmelCase(t *testing.T) {
+	t.Parallel()
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/amadeus/book_flight_normalize_test.go:14]: Fixed in 603233a — renamed to `TestBookFlightNormalize_SnakeCaseToCamelCase`.

File: connectors/amadeus/book_flight_normalize_test.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+		"idempotency_key": "key-1"
+	}`)
+
+	result := action.Normalize(input)
+
+	var m map[string]json.RawMessage
+	if err := json.Unmarshal(result, &m); err != nil {
+		t.Fatalf("unmarshal result: %v", err)
+	}
+
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/amadeus/book_flight_normalize_test.go:226]: Added the three deletion assertions in 603233a. They now verify `date_of_birth`, `first_name`, and `last_name` are removed from the output when their camelCase equivalents are present.

File: api/agent_approvals.go [hunk 6]
```diff
@@ -150,6 +150,21 @@ func handleAgentRequestApproval(deps *Deps) http.HandlerFunc {
...
+				if normalizer, ok := action.(connectors.Normalizer); ok {
+					if rawParams, hasParams := actionObj["parameters"]; hasParams {
+						actionObj["parameters"] = normalizer.Normalize(rawParams)
+						if updated, err := json.Marshal(actionObj); err == nil {
+							req.Action = updated
+						}
+					}
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply api/agent_approvals.go]: Fixed in 6968d2b. Merged both normalization blocks under a single GetAction lookup, consistent with action_execute.go.

File: connectors/amadeus/book_flight_normalize.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+						nameChanged = true
+					}
+					continue
+				}
+				if val, hasSnake := name[snake]; hasSnake {
+					name[camel] = val
+					delete(name, snake)
+					nameChanged = true
+				}
+			}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/amadeus/book_flight_normalize.go]: Good point. Split into `snakeToCamelTravelerFields` (date_of_birth only) and `snakeToCamelNameFields` (first_name, last_name) in 6968d2b. The name sub-object loop now only checks name-relevant fields.

File: connectors/amadeus/book_flight_normalize.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+				}
+				continue
+			}
+			if val, hasSnake := t[snake]; hasSnake {
+				t[camel] = val
+				delete(t, snake)
+				changed = true
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/amadeus/book_flight_normalize.go:96]: Good catch — removed the top-level name-field normalization in c0b0ca9. Name fields are now only normalized inside the `name` sub-object where they belong per the Amadeus schema. The top-level loop only handles `date_of_birth` → `dateOfBirth`.

**Comment by chiedoclaude** [comment]
## 🤖 Watch Session Summary

### Changes Made
No changes made during this session.

### Decision Log

#### ✅ Implemented
- greptile-apps[bot] asked for Delete snake_case keys when camelCase already present (top-level normalization) → implemented in `603233a`
- greptile-apps[bot] asked for Fix same stale-key bug in name sub-object normalization → implemented in `603233a`
- greptile-apps[bot] asked for Merge redundant GetAction lookups in action_execute.go → implemented in `603233a`
- greptile-apps[bot] asked for Fix test function name typo: CalmelCase → CamelCase → implemented in `603233a`
- greptile-apps[bot] asked for Add missing deletion assertions in CamelCaseTakesPrecedence test → implemented in `603233a`
- greptile-apps[bot] asked for Merge redundant GetAction lookups in agent_approvals.go → implemented in `6968d2b`
- greptile-apps[bot] asked for Split shared alias map into separate traveler and name field maps → implemented in `6968d2b`
- greptile-apps[bot] asked for Remove top-level name-field normalization to avoid leaving name fields outside the name sub-object → implemented in `c0b0ca9`

---
*Watch session ended after 6 minutes of inactivity. Processed 8 comments across 6 poll cycles.*

**Comment by chiedoclaude** [comment]
⚠️ **Auto-merge failed.** The `--automerge` flag was set, but the merge could not be completed. Please merge manually.
